common/app/views/support/cleaner/AmpAdCleaner.scala

```diff
@@ -12,84 +12,80 @@ import scala.xml.Elem
+  val IMG_BUFFER_FWD = 300 // really any non-p element type
+  val IMG_BUFFER_BWD = 200
+
+  /**
+    * Find ad slots
+    *
+    * Returns a list of elements *after* which ads should be directly
+    * placed.
+    *
+    * Ads are placed:
+    *
+    * * sufficiently far from other ads (MIN_CHAR_BUFFER characters away)
+    * * sufficiently far from non-text (p) elements (IMG_BUFFER_[FWD|BWD])
+    * * non adjacent to small (SMALL_PARA_CHARS) paragraphs
+    *
+    * These tests apply forwards and backwards, though in the case of
+    * non-text buffers, the values differ.
+    *
+    * Where the above tests are met, we say there is adequate 'buffer'
+    * around the ad and it can be placed.
+    */
+  def findAdSlots(elems: Vector[Element]): List[Element] = {
+    def isPara(elem: Element): Boolean = elem.tagName() == "p"
+    def suitableAdNeighbour(elem: Element): Boolean = isPara(elem) && elem.text.length > SMALL_PARA_CHARS
+
+    def hasForwardBuffer(elems: Vector[Element], index: Int): Boolean = {
+      val enoughCharsFwd = {
+        val fwdElems = elems.drop(index + 1)
+        val meetsThreshold = fwdElems.takeWhile(_.tagName == "p").map(_.text.length).sum >= IMG_BUFFER_FWD
+        meetsThreshold || !fwdElems.exists(_.tagName != "p")
       }
-    }
 
-    // propagate the constraints across the paragraphs forwards and then backwards
-    val propagatedConstraints = constraints.foldLeft((Nil: List[AdRepel], CHARS_BETWEEN_ADS)){ case ((accu, charsTilAd), currentElement) =>
+      val neighbourSuitable = elems.lift(index + 1).exists(suitableAdNeighbour)
 
-      val newCharsTilAd = charsTilAd - currentElement.length
-      def carryForwardRepel = currentElement.after < newCharsTilAd
+      enoughCharsFwd && neighbourSuitable
+    }
 
-      if (carryForwardRepel) {
-        (currentElement.copy(after = newCharsTilAd) :: accu, newCharsTilAd)
-      } else {
-        (currentElement :: accu, currentElement.after)
+    def hasBackwardBuffer(elems: Vector[Element], index: Int, textSinceLastAd: Int): Boolean = {
+      val enoughCharsBwd = {
+        val bwdElems = elems.take(index + 1).reverse // include element itself as ad will be placed after
+        val meetsThreshold = bwdElems.takeWhile(_.tagName == "p").map(_.text.length).sum >= IMG_BUFFER_BWD
+        meetsThreshold || !elems.exists(_.tagName != "p")
       }
 
-    }._1.foldLeft((Nil: List[AdRepel], 0)){ case ((accu, charsTilAd), currentElement) =>
-
-      val newCharsTilAd = charsTilAd - currentElement.length
-      def carryBackwardRepel = currentElement.before < newCharsTilAd
-
-      if (carryBackwardRepel) {
-        (currentElement.copy(before = newCharsTilAd) :: accu, newCharsTilAd)
-      } else {
-        (currentElement :: accu, currentElement.before)
-      }
+      suitableAdNeighbour(elems(index)) && textSinceLastAd >= MIN_CHAR_BUFFER && enoughCharsBwd
+    }
 
-    }._1
+    var charsScannedSinceLastAd = 0
+    var adSlots = List[Element]()
+    var adCount = 0
```

Comment:
I'd never considered using some vars in the original implementation, that's why I ended up using folds.  Since the code will go away soon anyway maybe it's fine to do that (plus I'm not sure what the dotcom coding standards are like), but personally in the general case I think vars are a line in the sand it's worth sticking to.  Otherwise we have to make a judgement in every case about whether the scope of the var is narrow enough to be understandable in one place.  The naming is certainly an improvement though!

Reply:
Interesting. I could definitely lift the vars into parameters of an inner function and go recursive.

Vars is definitely debatable. In general I agree that they are to be avoided, but within a function (and provided they don't leak) I think they are sometimes (rarely) appropriate.